File: client/src/app/classes/game-saved.ts
import { Player } from '@app/classes/player';
import { Spectator } from '@app/classes/spectator';

export class GameSaved {
    roomName: string;
    players: Player[];
    spectators?: Spectator[];
    winners?: Player[];
    numberOfTurns: number;
    gameStartDate: string;
    playingTime: string;
    nbLetterReserve: number;
    mapLettersOnStand: Map<string, string>;
    _id?: string;

    constructor(
        players: Player[],
        roomName: string,
        numberOfTurns: number,
        gameStartDate: string,
        playingTime: string,
        nbLetterReserve: number,
        spectators?: Spectator[],
        winners?: Player[],
    ) {
        this.roomName = roomName;
        this.players = players;
        this.spectators = spectators;
        this.winners = winners;
        this.numberOfTurns = numberOfTurns;
        this.playingTime = playingTime;
        this.nbLetterReserve = nbLetterReserve;
        this.gameStartDate = gameStartDate;
        this.mapLettersOnStand = new Map<string, string>();

        this.populateMap(this.players);
    }

    populateMap(players: Player[]) {
        for (const player of players) {
            this.mapLettersOnStand.set(player.name, this.lettersOnStand(player));
        }
    }

    lettersOnStand(player: Player): string {
        const listLetterStillOnStand: string[] = new Array<string>();
        for (const tile of player.stand) {
            if (tile.letter.value !== '') {
                listLetterStillOnStand.push(tile.letter.value);
            }
        }
        return listLetterStillOnStand.toString();
    }
}

File: client_leger/lib/models/game-saved.dart
import 'dart:core';
import 'package:client_leger/models/player.dart';
import 'package:client_leger/models/spectator.dart';

class GameSaved {

  // fields for the game-saved model used in database
  late String roomName;
  late List<String> players;
  late List<String> spectators;
  late List<String> winners;
  late List<num> scores;
  late num numberOfTurns;
  late String gameStartDate;
  late String playingTime;
  late num nbLetterReserve;
  late Map<String, String> mapLetterOnStand;
  late String? _id;

  GameSaved(List<Player> players,  this.roomName, this.numberOfTurns, this.playingTime, this.nbLetterReserve
      , this.gameStartDate, List<Spectator>? spectators, List<Player>? winners)
  {
    mapLetterOnStand = {};
    this.players = [];
    this.spectators = [];
    this.winners = [];
    scores = [];

    populateArrays(players, spectators, winners);
    populateMap(players);
  }

  GameSaved.fromJson(game) {
    roomName = game["roomName"];
    numberOfTurns = game["numberOfTurns"];
    playingTime = game["playingTime"];
    nbLetterReserve = game["nbLetterReserve"];
    gameStartDate = game["gameStartDate"];
    var playersString = game["players"];
    players = [];
    for (var pl in playersString){
        players.add(pl);
    }
    var winnersString = game["winners"];
    winners = [];
    for (var wn in winnersString) {
        winners.add(wn);
    }
    var spectatorsString = game["spectators"];
    spectators = [];
    for (var spec in spectatorsString) {
        spectators.add(spec);
    }
    var scoresString = game["scores"];
    scores = [];
    for (var sc in scoresString) {
        scores.add(sc);
    }
  }

  void populateArrays(List<Player> players, List<Spectator>? spectators, List<Player>? winners){
    for (var i = 0; i < players.length; i++) {
      this.players[i] = players[i].name;
    }
    for (var index = 0; index < 4; index++) {
      scores[index] = players[index].score;
    }
    if (spectators != null) {
      for (var index = 0; index < spectators.length; index++) {
        this.spectators[index] = spectators[index].name;
      }
    }
    if (winners != null) {
      for (var index = 0; index < winners.length; index++) {
        this.winners[index] = winners[index].name;
      }
    }
  }

  void populateMap(List<Player> players) {
    for (var player in players) {
      final entry = {player.name: lettersOnStand(player)};
      mapLetterOnStand.addEntries(entry.entries);
    }
  }

  String lettersOnStand(Player player) {
    const List<String> listLetterStillOnStand = [];
    for (var tile in player.stand) {
      if (tile.letter.value != '') {
        listLetterStillOnStand.add(tile.letter.value);
      }
    }
    return listLetterStillOnStand.toString();
  }

}
